File: u3-div-and-conq/mergesort.py
# mergesort program 
'''
It sorts a given array A by dividing it into two halves
A[0..[n/2]-1] and A[[n/2]..n - 1], sorting each of them recursively, and then
merging the two smaller sorted arrays into a single sorted one
'''


def mergesort(list):
    if len(list)==1:
        return list
    else:
        b=mergesort(list[:len(list)//2])
        c=mergesort(list[len(list)//2:])
        return merge(b,c,list)


def merge(b,c,a):
    i=0
    j=0
    k=0
    while i<len(b) and j<len(c):
        if b[i]<c[j]:
            a[k]=b[i]
            i+=1
        else:
            a[k]=c[j]
            j+=1
        k+=1
    #in case c is empty    
    while i<len(b):
        a[k]=b[i]
        i+=1
        k+=1
    #in case b is empty    
    while j<len(c):
        a[k]=c[j]
        j+=1
        k+=1
    return a 

def main():
    while 1:
        
        opt =int(input("Enter you choice: 1)mergeosrt 2) quit"))  
        if opt==1:
            n=int(input("Enter the number of elements in the list: "))
            list=[]
            if n>0:
                for i in range(n):
                    list.append(int(input("Enter the element: ")))      
                print("The sorted list is: ",mergesort(list))
            else:
                print("Invalid input")
        elif opt==2:
            break
        else:
            print("Invalid input")
main()            

File: u3-div-and-conq/mergesort.c
// mergesort program
#include<stdio.h>
#include<stdlib.h>

int main()
{
    int i,j,k,n,*a,*b,*c;
    printf("Enter the number of elements\n");
    scanf("%d",&n);
    a=(int*)malloc(n*sizeof(int));
    b=(int*)malloc(n*sizeof(int));
    c=(int*)malloc(n*sizeof(int));
    printf("Enter the elements\n");
    input_array(a,n);
    mergesort(a,b,c,0,n-1);
    printf("The sorted array is\n");
    output_array(a,n);
    return 0;
}
void input_array(int *a, int n)
{
    int i;
    for(i=0;i<n;i++)
    {
        scanf("%d",&a[i]);
    }
}
void output_array(int *a, int n)
{
    int i;
    for(i=0;i<n;i++)
    {
        printf("%d\n",a[i]);
    }
}
void mergesort(int *a, int *b, int *c, int l, int r)
{
    int m;
    if(l<r)
    {
        m=(l+r)/2;
        mergesort(a,b,c,l,m);
        mergesort(a,b,c,m+1,r);
        merge(a,b,c,l,m,r);
    }
}
void merge(int *a, int *b, int *c, int l, int m, int r)
{
    int i,j,k;
    i=l;
    j=m+1;
    k=l;
    while(i<=m && j<=r)
    {
        if(a[i]<=a[j])
        {
            c[k]=a[i];
            i++;
        }
        else
        {
            c[k]=a[j];
            j++;
        }
        k++;
    }
    while(i<=m)
    {
        c[k]=a[i];
        i++;
        k++;
    }
    while(j<=r)
    {
        c[k]=a[j];
        j++;
        k++;
    }
    for(i=l;i<=r;i++)
    {
        a[i]=c[i];
    }
}


